core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/latency/VMOperationRule.java

```diff
@@ -92,36 +98,112 @@ private Result evaluate(IItemCollection items, IQuantity warningLimit) {
-			return new Result(this, score,
-					MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_WARN), peakDuration),
-					MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_WARN_LONG),
-							peakDuration, operation, Encode.forHtml(caller.getThreadName()), timeStr),
+			String longMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_WARN_LONG_COMBINED_DURATION
+					: Messages.VMOperationRuleFactory_TEXT_WARN_LONG;
+			String shortMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_WARN_COMBINED_DURATION
+					: Messages.VMOperationRuleFactory_TEXT_WARN;
+			return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration),
+					MessageFormat.format(Messages.getString(longMessage), peakDuration, operation, caller, timeStr),
 					JdkQueries.VM_OPERATIONS_BLOCKING);
 		}
-		return new Result(this, score,
-				MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_OK), peakDuration), null,
+		String shortMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_OK_COMBINED_DURATION
+				: Messages.VMOperationRuleFactory_TEXT_OK;
+		return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration), null,
 				JdkQueries.FILE_READ);
 	}
 
+	private void findLongestEventInfo(IItemCollection items) {
+		List<IItem> sortedEvents = sortEventsByStartTime(items);
+		Iterator<IItem> i = sortedEvents.iterator();
+		IItem e1, e2 = null;
+		IItem startingEvent, curStartingEvent = null;
+		IQuantity longestDuration, curDuration = null;
+
+		if (i.hasNext()) {
+			e1 = i.next();
+			curStartingEvent = e1;
+			startingEvent = curStartingEvent;
+			curDuration = getDuration(e1);
+			longestDuration = curDuration;
+		} else {
+			return;
+		}
+		while (i.hasNext()) {
+			e2 = i.next();
+			double secsBetweenEvents = getStartTime(e2).subtract(getEndTime(e1)).doubleValueIn(UnitLookup.SECOND);
+			if (getCaller(e1).equals(getCaller(e2)) && getOperation(e1).equals(getOperation(e2))
+					&& secsBetweenEvents <= MAX_SECONDS_BETWEEN_EVENTS) {
+				curDuration = curDuration.add(getDuration(e2));
+			} else {
+				curDuration = getDuration(e2);
+				curStartingEvent = e2;
+			}
+
+			if (longestDuration.compareTo(curDuration) < 0) {
+				longestDuration = curDuration;
+				startingEvent = curStartingEvent;
+			}
+			e1 = e2;
+		}
+		this.longestDuration = longestDuration;
+		this.startingEvent = startingEvent;
+	}
+
+	private List<IItem> sortEventsByStartTime(IItemCollection items) {
+		List<IItem> sortedEvents = new ArrayList<>();
+		for (IItemIterable iter : items) {
+			for (IItem event : iter) {
+				sortedEvents.add(event);
+			}
+		}
+		Collections.sort(sortedEvents, new Comparator<IItem>() {
+			@Override
+			public int compare(IItem e1, IItem e2) {
+				return getStartTime(e1).compareTo(getStartTime(e2));
+			}
+		});
+		return sortedEvents;
+	}
```

Comment:
This is fairly expensive and unnecessary, during parsing events are stored in lanes by type that are disjoint (for events with durations) and sorted. You could use that fact and check for the time between two events in the same lane instead, possibly with a second pass over the IItemCollection if there are multiple lanes.

Reply:
If there are multiple lanes for one event type, even though the lanes themselves are sorted,  is it guaranteed that two consecutive events within a lane are actually consecutive in real time? For example if event A and B occur one after the other, is it not possible for A to be in a different lane than B and not necessarily be the first event that occurs in its lane or the last event that finishes in its lane, but be somewhere in the middle?

Reply:
The only guarantee is that the lanes are sorted on event times and are disjoint. So that might be a possible outcome.

Reply:
Ok,  I have left the sorting of events in my most recent commit https://github.com/openjdk/jmc/pull/42/commits/19fb291dbc838aeb583834690cce663f3a114f91 to deal with the case mentioned above. Although, yes, it is expensive and unnecessary to sort the events if there is only one lane, if there are multiple lanes, this sorting needs to occur to ensure that the events are being compared in the order the occur in real time.  

Let me know what you think.

Reply:
I think this might be alright as long as the amount of VM Operation events are fairly low (< 1000) but if there are too many then this approach would be quite slow. At that point we could file a bug to improve the performance here.